### PR TITLE
disable present bmc admin user and create new superuser

### PR DESCRIPTION
**Pull request by kolsa**

### CODE REVIEW

File: cmd/register/register.go [hunk 1]
```diff
@@ -210,9 +210,15 @@ func readIPMIDetails(eth0Mac string, hal hal.InBand) (*models.ModelsV1MachineIPM
 		if bmc == nil {
 			return nil, errors.New("unable to read ipmi bmc info configuration")
 		}
+
+		err := hal.BMCChangePassword(hal.BMCSuperUser(), "pw") //TODO pw should be provided via cmdline, for example
```
Comment (majst01) [inline cmd/register/register.go:214]: So the password must then be provided by a kernel commandline ?

Comment (kolsa) [reply cmd/register/register.go:214]: Or we introduce a grpc endpoint for this

**Comment by kolsa** [comment]
The last commit also introduce a new grpc client that encapsulates the lower level functions to securely communicate with metal-api via grpc, and provides a higher level API for the wait-for-allocation feature and the new introduced supermetal password retrieving endpoint

File: cmd/root.go [hunk 2]
```diff
@@ -134,6 +140,13 @@ func Run(spec *Specification, hal hal.InBand) (*event.EventEmitter, error) {
 		return eventEmitter, err
 	}
 
+	grpcClient, err := NewGrpcClient(certsClient, eventEmitter)
+	if err != nil {
+		log.Error("failed to fetch GRPC certificates", "error", err)
+		return eventEmitter, err
+	}
+	hammer.GrpcClient = grpcClient
```
Comment (majst01) [inline cmd/root.go:148]: grpcclient is required for UpdateBmcSuperuserPassword, so this is too late

Comment (kolsa) [reply cmd/root.go]: good catch :+1:

File: cmd/register/register.go [hunk 3]
```diff
@@ -202,28 +202,24 @@ func readIPMIDetails(eth0Mac string, hal hal.InBand) (*models.ModelsV1MachineIPM
 	details := &models.ModelsV1MachineIPMI{
 		Interface: &intf,
 	}
-	bmcversion := "unknown"
-	if hal.BMCPresent() {
+	bmcVersion := "unknown"
+	bmcConn := hal.BMCConnection()
+	if bmcConn.Present() {
```
Comment (majst01) [inline cmd/register/register.go:207]:
```suggestion
        // TODO further optimization in go-hal can use the vagrant detection to fill the register struct here
	if bmcConn.Present() {
```

File: cmd/root.go [hunk 4]
```diff
@@ -108,6 +107,27 @@ func Run(spec *Specification, hal hal.InBand) (*event.EventEmitter, error) {
...
+
+	skip, err := hammer.CreateBmcSuperuser()
+	if err != nil {
+		log.Error("failed to update bmc superuser password", "error", err)
+		return eventEmitter, err
+	}
+	if !skip {
```
Comment (majst01) [inline cmd/root.go:122]: I would suggest to put the logic to disable the existing superuser inside the "CreateBMCSuperuser" func

File: cmd/supwd.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+		return nil, nil, err
+	}
+	return v1.NewSuperUserPasswordClient(conn), conn, nil
+}
+
+// FetchSuperUserPassword tries to fetch the bmc superuser password from metla-api.
+// If no superuser password has been set in metal-api it returns an empty string and true as
+// the second return value, which indicates to skip further processing regarding the superuser password.
+// Otherwise that second return value is always false.
+func (c *GrpcClient) FetchSuperUserPassword() (string, bool, error) {
```
Comment (majst01) [inline cmd/supwd.go:23]: no need to export

**Review by majst01 — APPROVED** [review]

**Pull request merged**